### ISSUE TITLE
Neural network with several target vectors to run on TPU

### ISSUE THREAD

**Issue by Kirill94a** [tensorflow/tensorflow#37468]
I try to create my own neural network with several outputs (it follows that there are several target vectors); these target vectors dynamically change during training and depend on the predictions of the neural network. That is, it looks like the following: `y_proba = f(input), y_true = g(y_proba)`, where `input` - input vector, `f` - function of the neural network, `y_proba` - predicted target vector, `g` - function of changing the predicted target vector, `y_true` - target vector.
The problem is that I cannot implement this neural network using Keras and TPUEstimator in TensorFlow 2.1 to run on TPU.
I tried to solve this problem in two ways.

**1. The first way to solve the problem**

I changed the source code of Keras inside TensorFlow to run on CPU and TPU to solve this problem.

Changes to run on CPU in `tensorflow_core/python/keras/engine/training_arrays.py` inside `model_iteration` function:
```
#calling my own function to change ins_batch
batch_outs = f(ins_batch)
```

Changes to run on TPU in `tensorflow_core/python/keras/engine/training_eager.py` inside `_model_loss` function:
```
if targets:
  targets = training_utils.cast_if_floating_dtype_and_mismatch(targets, outs)
#calling my own function to change targets
```

Changes to run on TPU without experimental functions (setting `experimental_run_tf_function=False` when compiling the model as described in [this](https://github.com/tensorflow/tensorflow/issues/33729#issuecomment-562656807)) in `tensorflow_core/python/keras/engine/training_arrays.py` inside `model_iteration` function:
```
actual_inputs = ins()
#calling my own function to change actual_inputs
batch_outs = f(actual_inputs)
```

To run on CPU, I can dynamically change the target vector inside TensorFlow, and the neural network is trained. But adding this feature inside TensorFlow to run on TPU has many nuances and works like magic and I couldn’t train the neural network.

**2. The second way to solve the problem**
I tried to implement my neural network based on `TPUEstimator` using `RegressionHead` and `MultiHead`. To do this, I used examples ([mnasnet](https://github.com/tensorflow/tpu/tree/master/models/official/mnasnet) and [convolutional_network](https://github.com/aymericdamien/TensorFlow-Examples/blob/master/examples/3_NeuralNetworks/convolutional_network.py)).

This is the part of the program that implements my own neural network using `TPUEstimator`:

```
def get_model(features, input_shape, reuse):
    with tf.variable_scope('model', reuse=reuse):
        seqs = []
        n_filters = 8
        for i in range(n):
            seqs.append(get_conv2d(
                          seqs[-1].get_x(),
                          (3, 3, 8, n_filters),
                          reuse=reuse))
        outputs = []
        heads = []
        for x in seqs:
            outputs.append(OutputLayer(name="output_"+x.name)(x.get_x()))
            heads.append(tf.estimator.RegressionHead(
                                           label_dimension=32*32*8,
                                           name="head_"+x.name))
        head = tf.estimator.MultiHead(heads)
    return head, outputs

def model_fn(features, labels, mode, params):
    batch_size = 8 * params['batch_size']

    head, logits_train = get_model(features, params['input_shape'], reuse=False)
    logits_train_dic = {}
    for i in range(n):
        logits_train[i] = tf.reshape(logits_train[i], (batch_size, 32*32*8,))
        logits_train_dic["head_{}".format(i)] = logits_train[i]
    pred_classes = tf.argmax(logits_train, axis=1)
    if mode == tf.estimator.ModeKeys.PREDICT:
        return tf.estimator.tpu.TPUEstimatorSpec(mode, predictions=pred_classes)

    new_labels = {}
    for key in labels:
        new_labels[key] = labels[key][0]
    loss = 0.0
    for key in logits_train_dic:
        logit_train = logits_train_dic[key]
        loss += tf.square(labels[key]-logit_train)
    loss_op = tf.reduce_mean(loss)
    optimizer = tf.train.AdamOptimizer(learning_rate=params['learning_rate'])
    if params['use_tpu']:
        optimizer = tf.tpu.CrossShardOptimizer(optimizer)
    train_op_fn = lambda loss_op: optimizer.minimize(
                                  loss_op,
                                  global_step=tf.train.get_global_step())
                
    estim_specs = head.create_estimator_spec(
                  features={"x": features},
                  labels=new_labels,
                  mode=mode,
                  logits=logits_train_dic,
                  train_op_fn=train_op_fn)
    return estim_specs

model = tf.estimator.tpu.TPUEstimator(
          model_fn, use_tpu=True,
          config=config,
          train_batch_size=train_batch_size,
          params=params)
```
As I understand it, this way of implementing a neural network makes it possible to dynamically change labels (in particular, target vectors).
I implemented the neural network itself using `TPUEstimator`, but I faced with another problem - "Graph was finalized" is displayed, but the training of the neural network does not start.

How to implement a neural network with several dynamically changing target vectors, which depend on neural network predictions, on TPU?

**Comment by gadagashwini** [comment]
@Kirill94a, Will it be possible to post the complete standalone code to analyze the reported issue. Thanks!

**Comment by Kirill94a** [comment]
I found the cause of the problem - I made a error when connecting to TPU. Issue may be closed. Thank you!
